src/ejabberd_service.erl

```diff
@@ -440,24 +445,90 @@ fsm_limit_opts(Opts) ->
             end
     end.
 
+try_register_routes(StateData) ->
+    try_register_routes(StateData, 3).
+
+try_register_routes(StateData, Retries) ->
+    case register_routes(StateData) of
+        ok ->
+            send_text(StateData, <<"<handshake/>">>),
+            {next_state, stream_established, StateData};
+        {error, Reason} ->
+            RoutesInfo = lookup_routes(StateData),
+            ConflictBehaviour = StateData#state.conflict_behaviour,
+            ?ERROR_MSG("event=component_registration_conflict "
+                       "reason=~p conflict_behaviour=~p retries=~p routes_info=~p",
+                       [Reason, ConflictBehaviour, Retries, RoutesInfo]),
+            handle_registration_conflict(ConflictBehaviour, RoutesInfo, StateData, Retries)
+    end.
+
+routes_info_to_pids(RoutesInfo) ->
+    {_Hosts, ExtComponentsPerHost} = lists:unzip(RoutesInfo),
+    %% Flatten the list of lists
+    ExtComponents = lists:append(ExtComponentsPerHost),
+    Handlers = lists:map(fun external_component_to_handler/1, ExtComponents),
```

Comment:
I'm being picky but it's so overengineered `[ Handler || #external_component{ handler = Handler } <- ExtComponents ]`. :D